substrate-node/pallets/pallet-smart-contract/src/lib.rs

```diff
@@ -607,59 +843,178 @@ impl<T: Config> Pallet<T> {
+        Contracts::<T>::insert(&contract.contract_id, &contract);
+
+        Self::deposit_event(Event::ContractUpdated(contract));
+
+        Ok(().into())
+    }
+
+    #[transactional]
+    pub fn _create_deployment(
+        account_id: T::AccountId,
+        capacity_reservation_id: u64,
+        deployment_hash: DeploymentHash,
+        deployment_data: DeploymentDataInput<T>,
+        resources: Resources,
+        public_ips: u32,
+    ) -> DispatchResultWithPostInfo {
         let twin_id = pallet_tfgrid::TwinIdByAccountID::<T>::get(&account_id)
             .ok_or(Error::<T>::TwinNotExists)?;
-        let contract = Self::_create_contract(
+
+        let cr_contract = Contracts::<T>::get(capacity_reservation_id)
+            .ok_or(Error::<T>::CapacityReservationNotExists)?;
+        let capacity_reservation_contract = Self::get_capacity_reservation_contract(&cr_contract)?;
+        ensure!(
+            cr_contract.twin_id == twin_id,
+            Error::<T>::TwinNotAuthorized
+        );
+
+        // If the contract with hash and node id exists and it's in any other state then
+        // contractState::Deleted then we don't allow the creation of it.
+        // If it exists we allow the user to "restore" this contract
+        if ContractIDByNodeIDAndHash::<T>::contains_key(
+            capacity_reservation_contract.node_id,
+            &deployment_hash,
+        ) {
+            let contract_id = ContractIDByNodeIDAndHash::<T>::get(
+                capacity_reservation_contract.node_id,
+                &deployment_hash,
+            );
+            let contract = Contracts::<T>::get(contract_id).ok_or(Error::<T>::ContractNotExists)?;
+            if !contract.is_state_delete() {
+                return Err(Error::<T>::ContractIsNotUnique.into());
+            }
+        }
+
+        let public_ips_list: BoundedVec<
+            PublicIP<
+                <T as pallet_tfgrid::Config>::PublicIP,
+                <T as pallet_tfgrid::Config>::GatewayIP,
+            >,
+            MaxNodeContractPublicIPs<T>,
+        > = vec![].try_into().unwrap();
+
+        // Get the Contract ID map and increment
+        let mut id = DeploymentID::<T>::get();
+        id = id + 1;
+
+        // Prepare Deployment struct
+        let mut deployment = types::Deployment {
+            id,
             twin_id,
-            types::ContractData::RentContract(types::RentContract { node_id }),
-            solution_provider_id,
+            capacity_reservation_id,
+            deployment_hash: deployment_hash.clone(),
+            deployment_data,
+            public_ips,
+            public_ips_list,
+            resources,
+        };
+
+        Self::_reserve_ip(id, &mut deployment)?;
+        Self::_add_deployment_to_capacity_reservation_contract(capacity_reservation_id, id)?;
+        Self::_change_used_resources_on_capacity_reservation(
+            capacity_reservation_id,
+            Resources::empty(),
+            resources,
```

Comment:
Maybe it's something that does not really matter on performance but in these 2 function the capacity contract is fetched, mutated and inserted. Maybe you could fetch and pass a mutable reference and save when you are done.